util/canonical-serializer/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+//! Canonical Serializer
+//! see Readme for details.
+//!
+
+use bytes::Bytes;
+use failure::{ensure, Error};
+use numext_fixed_hash::H256;
+use std::io::Write;
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+pub trait CanonicalSerialize {
+    fn serialize<W: Write>(&self, serializer: &mut CanonicalSerializer<W>) -> Result<()>;
+}
+
+pub struct CanonicalSerializer<W> {
+    buf: W,
+}
+
+impl<W: Write> CanonicalSerializer<W> {
+    pub fn new(buf: W) -> Self {
+        CanonicalSerializer { buf }
+    }
+
+    pub fn encode_u8(&mut self, v: u8) -> Result<&mut Self> {
+        self.buf.write_all(&[v])?;
+        Ok(self)
+    }
+
+    pub fn encode_u32(&mut self, v: u32) -> Result<&mut Self> {
+        self.buf.write_all(&v.to_le_bytes())?;
+        Ok(self)
+    }
+
+    pub fn encode_u64(&mut self, v: u64) -> Result<&mut Self> {
+        self.buf.write_all(&v.to_le_bytes())?;
+        Ok(self)
+    }
+
+    pub fn encode_h160(&mut self, v: &[u8]) -> Result<&mut Self> {
+        ensure!(
+            v.len() == 20,
+            "serialize H160 length error expect 20, got {}",
+            v.len()
+        );
+        self.buf.write_all(v)?;
+        Ok(self)
+    }
+
+    pub fn encode_h256(&mut self, v: &[u8]) -> Result<&mut Self> {
+        ensure!(
+            v.len() == 32,
+            "serialize H256 length error expect 32, got {}",
+            v.len()
+        );
+        self.buf.write_all(v)?;
+        Ok(self)
+    }
+
+    pub fn encode_u256(&mut self, v: &[u8]) -> Result<&mut Self> {
+        ensure!(
+            v.len() == 32,
+            "serialize U256 length error expect 32, got {}",
+            v.len()
+        );
+        self.buf.write_all(v)?;
+        Ok(self)
+    }
+
+    pub fn encode_fix_length_bytes(&mut self, v: &[u8], len: usize) -> Result<&mut Self> {
```

Comment:
So, I have to call `v.len()` before call this method, then this method call `v.len()` again?

Reply:
no, this function is only for the fixed length data, like a proposal_id, the length is always 10

Reply:
So, you should not test `v.len() == len`, just test `v.len() >= len` then `write_all(&v[0..len])`.

Reply:
No, the purpose is you should not use a variable length data with this method.

In your case the length of v is longer than len, this is usually misused.

For example to encode an H256 `encode_fix_length_bytes(&h256, 32)`, the `len` variable should be a constant for a specific type data.

Reply:
- If they input a fixed bytes, why they have to input the length?

- The input `len` is only for test, and it's **implicit**, if the users do the test before call this, this test will be **redundant**.

  And in most scenarios, they have to do the same test before call this.

  I have to call `writer.encode_fix_length_bytes(slice[..32], 32)`, if the fixed bytes was included in a larger bytes, then it will test `32 == 32` again, no meaningless.

- The input `len` doesn't bring any features.

- IMHO:

  - Don't input the `len`, force users who call it to do tests.

  - Or, just test `v.len() >= len` then `write_all(&v[0..len])`, let the `len` do something, make it be more flexible.

  - Or, `fn encode_fix_length_bytes(&mut self, v: T, len: usize) where T: AsRef<[u8]>;`.

    I think this is not suitable for this API.

Reply:
The third one makes sense. It is too easy to write errors without the assert.